### PR TITLE
fix: fail Gate C closed when an external originator has no resolved tier (#1059)

### LINKED ISSUE #1059: Gate C fails open when an external originator has no tier stamped
## Context

While reviewing the Gate C escalation-judge work in #1057 (closes #950), both an independent security review and CodeRabbit flagged the same fail-open in `getInitiatingTier()` (`src/contacts/principal.ts`). It was deliberately deferred out of #1057 to avoid a behavior change mid-migration; this issue tracks closing it.

Related discussion: https://github.com/josephfung/curia/pull/1057#discussion_r3442071995

## Problem

`getInitiatingTier(metadata)` returns `null` for three distinct cases, and Gate C is skipped (fail-open) whenever it returns `null`:

1. **system**-originated tasks — genuinely no external contact → safe to skip.
2. **agent**-originated tasks — genuinely no external contact → safe to skip.
3. **external** originator that carries no `tier` field (a pre-#950 task in flight during a deploy, or a dispatcher stamping defect) → **a consequential external action bypasses tier enforcement entirely**, including the escalation judge.

Cases 1–2 are correct. Case 3 is a fail-open: an externally-initiated, consequential action (e.g. `email-send` / `signal-send`) auto-executes with no tier gate. Now that #1057 makes Gate C the authoritative third-party-facing enforcement point, a `null`-tier fall-through defeats all of it.

The current behavior is also **silent** — the skip logs nothing, so a stamping defect would be invisible.

## Proposed approach

Distinguish "structurally not externally-initiated" from "external contact with a missing tier":

- `systemRole === 'system' | 'agent'` → skip Gate C (unchanged, safe).
- External originator (any other `systemRole`) with `tier == null` → **fail closed** (escalate/block), not skip.
- Emit a `warn` log on the fail-closed path (skill name + `action_risk`) so an unstamped-tier external task is observable rather than silent.

Note the migration consideration: confirm the dispatcher now always stamps `tier` for external originators before flipping this, so legitimate in-flight tasks don't all start escalating. If a backward-compat window is still needed, gate the fail-closed behavior so only genuinely external-with-missing-tier tasks escalate.

## Acceptance criteria

- [ ] An externally-originated task (`systemRole` not `system`/`agent`) with a missing/`null` `tier` and a non-`none` `action_risk` **fails closed** (Gate C escalates; the skill handler does not execute).
- [ ] `system`-originated and `agent`-originated tasks still skip Gate C (no regression).
- [ ] A task with no originator at all is handled explicitly (decide skip vs. fail-closed and document it; default to fail-closed for consequential actions if an external origin can't be ruled out).
- [ ] The fail-closed path emits a `warn` log including skill name and `action_risk`.
- [ ] The two `execution.test.ts` tests currently asserting fail-open for the external-missing-tier case are inverted to assert fail-closed; explicit system/agent skip tests remain.
- [ ] `pnpm run typecheck` and the unit suite pass.

## Before implementing: measure real-world incidence first

A near-term observability log shipped with #1057 (commit `22ab8408`): when an **external** originator with no resolved tier drives a consequential action, Gate C is skipped (still fail-open) but now emits:

```
autonomy gate: Gate C skipped — external originator has no resolved tier (fail-open, see #1059)
```

(structured `warn`, with `skillName`, `actionRisk`, `originator`, `agentId`, `taskEventId`.)

**Do not flip to fail-closed blind.** First query the audit/observability logs **across instances** for this message to understand:
- How often the path is actually hit (per instance, over a representative window).
- Which originators and skills trigger it — i.e. is it only pre-#950 in-flight tasks (transient, drains after deploy) or a live dispatcher stamping gap (persistent)?
- Whether flipping to fail-closed would escalate any legitimate flow that currently auto-runs.

Use that evidence to choose the rollout: a direct flip if incidence is ~zero / purely transient, or a staged approach (warn-only window → fail-closed) if real traffic is affected. Record the findings in this issue before implementing.

## Additional acceptance criterion

- [ ] The audit/observability logs have been checked across instances for occurrences of the fail-open `warn`, and the findings (frequency, affected originators/skills, transient vs. persistent) are recorded in this issue and used to justify the chosen rollout (direct flip vs. staged).

### PR DESCRIPTION
**Pull request by josephfung**
## Summary

Closes #1059.

Gate C (the contact-tier enforcement gate in `src/skills/execution.ts`) previously **failed open** whenever `getInitiatingTier()` returned `null`. That `null` collapses three distinct cases — and one of them is an **external** contact whose `tier` was never stamped (a pre-#950 task in flight during a deploy, or a dispatcher stamping defect). For that case, a consequential action (`email-send`, `signal-send`, …) auto-executed with **no tier gate at all**, defeating the escalation judge wired in #1057.

This change distinguishes *structurally-not-external* from *external-with-missing-tier*:

- `systemRole === 'system' | 'agent'` → **skip** (unchanged, safe — no external contact).
- Task with **no originator at all** → **skip** (structurally internal; see decision note below).
- **External** originator (`systemRole` not system/agent) with `tier == null` and `action_risk !== 'none'` → **fail closed**: escalate via the existing tier-gate error path (creates an approval request when wired) and the handler does not run.
- `action_risk === 'none'` → still allowed (no external effect to gate).

The fail-closed path emits a structured `warn` (`skillName`, `actionRisk`, `originatorContactId`, full `originator`, `agentId`, `taskEventId`) so an unstamped-tier external task is observable rather than silent.

## "No originator" decision (documented per the issue's AC)

A task with no originator continues to **skip** Gate C. The only `invoke` callers are the agent runtime (always threads task metadata) and the checkpoint processor, which invokes `extract-facts`/`extract-relationships` (`low` risk) with **no** metadata. Failing closed on no-originator would block legitimate internal KG writes with zero security benefit. "Metadata present but no originator" is also not reliably external — the bullpen dispatcher creates internal inter-agent tasks that omit the originator — so an absent originator is treated as structurally internal at this layer.

## Residual fail-open (pre-existing, out of scope, follow-up filed)

Pre-PR review surfaced that the dispatcher only stamps an originator when `senderContext.resolved` is true, so a **truly unresolved external sender** dispatches a task with metadata but no originator and reaches Gate C as "no originator" → skipped. This is pre-existing and can only be fixed **upstream at the dispatcher** (Gate C can't tell an unresolved sender from an internal bullpen task). Tracked separately.

## Rollout

#1057 only just started deploying and is not yet in production, so there's no live incidence data to query (per the issue's measurement step). The path is reachable only by transient pre-#950 in-flight tasks (drain after deploy) or a stamping defect (which we want surfaced, not tolerated), so a direct flip is correct; a warn-only staging window would only delay closing a known hole.

## Testing

- `pnpm run typecheck` — clean.
- `src/skills/execution.test.ts` + `src/contacts/principal.test.ts` — 86 pass. Inverted the external-missing-tier fail-open test to assert fail-closed; added none-risk-allowed, warn-emission, and no-originator-skip tests; retained explicit system/agent skip tests.
- Full unit suite passes (one DB-integration test mislocated under `tests/unit/` requires `DATABASE_URL` and is unrelated — flagged below).

## Notes / tech debt spotted (not touched)

- `tests/unit/autonomy/autonomy-service-pagination.test.ts` requires a live `DATABASE_URL` despite living under `tests/unit/` — it's effectively an integration test and fails without Postgres. Worth relocating to `tests/integration/`.
---

## Update: dispatcher defence-in-depth folded in

A deeper trace (prompted by review) surfaced that the *everyday* brand-new-sender case is already handled — the channel adapter pre-creates the sender's contact with `tier='unknown'` ([email-adapter.ts:434-440](src/channels/email/email-adapter.ts#L434-L440)) before dispatch, so Gate C escalates correctly. But the Dispatch layer should not *assume* the Channel layer always does this: contact creation is skipped when the adapter's rate caps are hit, creation fails, or the From is malformed. In those degraded cases the sender reached dispatch unresolved → no originator → Gate C skipped.

So the dispatcher now **always stamps an originator** for inbound messages ([dispatcher.ts:668](src/dispatch/dispatcher.ts#L668)): resolved → real tier; unresolved → `{ contactId: senderId, systemRole: null, tier: 'unknown' }`. An `unknown`-tier originator flows through Gate C's *normal* path (which escalates consequential third-party actions), so the gap closes even when the channel layer didn't pre-create the contact. This also makes "no originator at all ⇒ structurally internal" finally airtight.

Reviewed for blast radius (always-present originator + non-UUID `ctx.caller.contactId`): the elevated-skill gate still blocks unresolved senders (`systemRole: null`), and every consumer of a non-UUID caller id is guarded (`UUID_RE` checks; the entity-context assembler catches Postgres `22P02`; audit columns are `TEXT`). No regressions.

New test: `tests/unit/dispatch/dispatcher.test.ts` — "stamps an unknown-tier originator for an unresolved inbound sender". Full unit suite green.

### CODE REVIEW

**Comment by josephfung** [comment]
@coderabbitai review

File: src/skills/execution.test.ts [hunk 1]
```diff
@@ -951,11 +954,68 @@ describe('autonomy gates', () => {
         },
       });
 
+      expect(result.success).toBe(false);
+      expect(handler.execute).not.toHaveBeenCalled();
+    });
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/skills/execution.test.ts:959]: Declining this one. These are Vitest tests that `await layer.invoke(...)` and then assert on the resolved result (`expect(result.success).toBe(false)`, etc.). The skill layer never throws — it returns `{ success, error }` — so there is nothing to catch here. If an unexpected rejection ever did occur, the correct behavior is for it to propagate and fail the test loudly; wrapping the `await` in `try/catch` would swallow exactly the failure signal the test exists to surface (and a `.catch()` that rethrows is a no-op). So the no-unhandled-async rule does not apply to assert-on-result test bodies.

File: src/skills/execution.test.ts [hunk 2]
```diff
@@ -951,11 +954,68 @@ describe('autonomy gates', () => {
...
+            channel: 'email',
+            initiatedAt: new Date().toISOString(),
+            // no tier
+          },
+        },
+      });
+
       expect(result.success).toBe(true);
       expect(handler.execute).toHaveBeenCalledOnce();
     });
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/skills/execution.test.ts:982]: Declining this one. These are Vitest tests that `await layer.invoke(...)` and then assert on the resolved result (`expect(result.success).toBe(false)`, etc.). The skill layer never throws — it returns `{ success, error }` — so there is nothing to catch here. If an unexpected rejection ever did occur, the correct behavior is for it to propagate and fail the test loudly; wrapping the `await` in `try/catch` would swallow exactly the failure signal the test exists to surface (and a `.catch()` that rethrows is a no-op). So the no-unhandled-async rule does not apply to assert-on-result test bodies.

File: src/skills/execution.test.ts [hunk 3]
```diff
@@ -951,11 +954,68 @@ describe('autonomy gates', () => {
...
+            // no tier
+          },
+        },
+      });
+
+      expect(warn).toHaveBeenCalledWith(
+        expect.objectContaining({ skillName: 'email-send', actionRisk: 'medium' }),
+        expect.stringContaining('failing closed'),
+      );
+    });
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/skills/execution.test.ts:1011]: Declining this one. These are Vitest tests that `await layer.invoke(...)` and then assert on the resolved result (`expect(result.success).toBe(false)`, etc.). The skill layer never throws — it returns `{ success, error }` — so there is nothing to catch here. If an unexpected rejection ever did occur, the correct behavior is for it to propagate and fail the test loudly; wrapping the `await` in `try/catch` would swallow exactly the failure signal the test exists to surface (and a `.catch()` that rethrows is a no-op). So the no-unhandled-async rule does not apply to assert-on-result test bodies.

File: src/skills/execution.test.ts [hunk 4]
```diff
@@ -951,11 +954,68 @@ describe('autonomy gates', () => {
...
+
+    it('skips Gate C when there is no originator in task metadata (structurally internal, #1059)', async () => {
```
[the remark this answers is not in the document]
Comment (josephfung) [reply src/skills/execution.test.ts:1013]: Declining this one. These are Vitest tests that `await layer.invoke(...)` and then assert on the resolved result (`expect(result.success).toBe(false)`, etc.). The skill layer never throws — it returns `{ success, error }` — so there is nothing to catch here. If an unexpected rejection ever did occur, the correct behavior is for it to propagate and fail the test loudly; wrapping the `await` in `try/catch` would swallow exactly the failure signal the test exists to surface (and a `.catch()` that rethrows is a no-op). So the no-unhandled-async rule does not apply to assert-on-result test bodies.

File: CHANGELOG.md [hunk 5]
```diff
@@ -29,6 +29,8 @@ bus event types) are noted explicitly even in the `0.x` range.
 
 ### Security
 
+- **Gate C fail-closed for unstamped external tiers** — an external originator with no resolved tier now escalates a consequential action instead of bypassing the tier gate; emits a `warn` so the case is observable. (#1059)
+- **Dispatcher stamps an `unknown`-tier originator for unresolved inbound senders** — defence in depth so Gate C enforces tier policy even when the channel layer's first-contact creation was skipped, rather than trusting the channel layer to always pre-create the contact. (#1059)
```
[the remark this answers is not in the document]
Comment (josephfung) [reply CHANGELOG.md]: Good catch, fixed in c8d42a71. Both Security bullets are now single-clause:

- **Gate C fail-closed for unstamped external tiers** — an external originator with no resolved tier now escalates instead of bypassing the gate. (#1059)
- **Dispatcher stamps an `unknown`-tier originator for unresolved inbound senders** — defence in depth so Gate C still enforces tier policy if the channel layer skipped contact creation. (#1059)

Dropped the `warn`/implementation detail to stay within the ~15-word target.
